src/libostree/ostree-sysroot.c

```diff
@@ -1645,22 +1653,52 @@ ostree_sysroot_simple_write_deployment (OstreeSysroot      *sysroot,
        * deployments, fall back on merge deployment */
       const gboolean passed_crossover = booted_deployment ? !before_booted : !before_merge;
 
+      /* Try extracting a version for this deployment. */
+      const gchar *csum = ostree_deployment_get_csum (deployment);
+      g_autoptr(GVariant) variant = NULL;
+      g_autoptr(GVariant) metadata = NULL;
+
+      g_autofree char *deployment_version = NULL;
+
+      OstreeRepo *repo = ostree_sysroot_repo (sysroot);
+      /* XXX Copying ot_admin_checksum_version() + bits from
+       *     ot-admin-builtin-status.c.  Maybe this should be
+       *     public API in libostree? */
+      if (ostree_repo_load_variant (repo, OSTREE_OBJECT_TYPE_COMMIT, csum,
+                                    &variant, NULL))
+        {
+          metadata = g_variant_get_child_value (variant, 0);
+          g_variant_lookup (metadata, OSTREE_COMMIT_META_KEY_VERSION, "s", &deployment_version);
+        }
+
+      if ((is_booted || is_merge) && !last_version)
+        last_version = g_steal_pointer (&deployment_version);
+
+      const gboolean is_previous_version =
+          (osname_matches && last_version && passed_crossover &&
+           (g_strcmp0 (deployment_version, last_version) < 0));
+
       /* Retain deployment if:
        *   - we're explicitly asked to, or
        *   - it's pinned
        *   - the deployment is for another osname, or
        *   - we're keeping pending deployments and this is a pending deployment, or
        *   - this is the merge or boot deployment, or
        *   - we're keeping rollback deployments and this is a rollback deployment
+       *   - we're keeping the previous version deployment
        */
       if (retain
           || ostree_deployment_is_pinned (deployment)
           || !osname_matches
           || (retain_pending && !passed_crossover)
           || (is_booted || is_merge)
-          || (retain_rollback && passed_crossover))
+          || (retain_rollback && passed_crossover)
+          || (retain_previous && is_previous_version && !n_previous))
         g_ptr_array_add (new_deployments, g_object_ref (deployment));
 
+      /* Just keep one previous version */
+      n_previous += (guint8)is_previous_version;
```

Comment:
Hmm, how about just a `retained_previous_version` boolean flag instead to detect that we've already done it?

Reply:
Ok